src/core/Database.cpp

```diff
@@ -470,30 +470,52 @@ Database* Database::unlockFromStdin(QString databaseFilename, QString keyFilenam
     return Database::openDatabaseFile(databaseFilename, compositeKey);
 }
 
-QString Database::saveToFile(QString filePath)
+/**
+ * Save the database to a file.
+ *
+ * This function uses QTemporaryFile instead of QSaveFile due to a bug
+ * in Qt (https://bugreports.qt.io/browse/QTBUG-57299) that may prevent
+ * the QSaveFile from renaming itself when using DropBox, Drive, or OneDrive.
+ *
+ * The risk in using QTemporaryFile is that the rename function is not atomic
+ * and may result in loss of data if there is a crash or power loss at the
+ * wrong moment.
+ *
+ * @param filePath Absolute path of the file to save
+ * @param keepOld Rename the original database file instead of deleting
+ * @return error string, if any
+ */
+QString Database::saveToFile(QString filePath, bool keepOld)
 {
     KeePass2Writer writer;
-    QSaveFile saveFile(filePath);
-    if (saveFile.open(QIODevice::WriteOnly)) {
-
+    QTemporaryFile saveFile;
+    if (saveFile.open()) {
         // write the database to the file
         setEmitModified(false);
         writer.writeDatabase(&saveFile, this);
         setEmitModified(true);
 
         if (writer.hasError()) {
+            // the writer failed
             return writer.errorString();
         }
 
-        if (saveFile.commit()) {
+        saveFile.close(); // flush to disk
+
+        if (keepOld) {
+            QFile::remove(filePath + ".old");
```

Comment:
.old should be inserted between the base filename and the .kdbx extension.

Reply:
Why you make this difficult? 😝

Reply:
😏